### PR TITLE
Workorder.lic improvements

### PR DESCRIPTION
**Pull request by Hazado**
Add @retain_crafting_materials to shaping, sewing, and knitting
Changes to stone carving to allow use of left over pieces after carving

### CODE REVIEW

File: workorders.lic [hunk 1]
```diff
@@ -221,7 +221,8 @@ class WorkOrders
   def carve_items(info, item, quantity)
     ensure_copper_on_hand(@cash_on_hand || 5000, @settings)
     recipe, items_per_stock, spare_stock, scrap = find_recipe(info, item, quantity)
-
+    rock_noun = %w[deed pebble stone rock rock boulder]
```
Comment (Sarvatt) [inline workorders.lic]: an extra rock here

Comment (Hazado) [reply workorders.lic]: Yes intentional
Deed, pebble, stone, small rock, large rock, boulder

File: workorders.lic [hunk 2]
```diff
@@ -246,6 +247,9 @@ class WorkOrders
     stow_hands
 
     quantity.times do |count|
+      if (info['stock-volume'] - (count * recipe['volume'])) > 0
```
Comment (rpherbig) [inline workorders.lic]: 💡 I'd like to see `(info['stock-volume'] - (count * recipe['volume'])` extracted into a well named var to give more context to the next dev that comes along

File: workorders.lic [hunk 3]
```diff
@@ -371,7 +385,9 @@ class WorkOrders
       end
     end
     leftover = (quantity * recipe['volume']) % 10 != 0
-    dispose("#{info['sew-stock-name']} cloth") if leftover
+    dispose("#{info['sew-stock-name']} cloth") if leftover &&  @retain_crafting_materials
```
Comment (rpherbig) [inline workorders.lic]: ❓ Should this be `!@retain_crafting_materials`?

File: workorders.lic [hunk 4]
```diff
@@ -246,6 +247,9 @@ class WorkOrders
     stow_hands
 
     quantity.times do |count|
+      if (info['stock-volume'] - (count * recipe['volume'])) > 0
+        rock_count = info['stock-volume'] - (count * recipe['volume'])
```
Comment (rpherbig) [inline workorders.lic]: 💡 `rock_noun[rock_count]` could also use a well named var

File: workorders.lic (hunk 1 above)
Comment (rpherbig) [reply workorders.lic]: ❓ I'm not clear on how `rock_noun` and `rock_count` work. Isn't `stock-volume` assuming you get a `stone`?

If I'm reading the loop on line 249 correctly, the first time through the loop `count` is zero, so `rock_count = info['stock-volume']` (i.e. buy a `rock`). How does that account for when `recipe['volume']` requires something bigger than a rock (or allows for something smaller like a pebble`?

Comment (Hazado) [reply workorders.lic]: No currently in crafting yaml its a small rock,
Volume 3. Currently no workorders in recipes are larger than 3 volume.

File: workorders.lic [hunk 5]
```diff
@@ -246,6 +247,7 @@ class WorkOrders
     stow_hands
 
     quantity.times do |count|
+      dispose("#{info['stock-name']} #{material_noun[material_volume+recipe['volume']]}") if (count > 0 && spare_stock)
```
Comment (rpherbig) [inline workorders.lic]: Thank you for renaming `material_[noun|volume]`, it makes this clearer.

Could you extract `#{material_noun[material_volume+recipe['volume']]}` into a well-named variable to help provide some context?

File: workorders.lic [hunk 6]
```diff
@@ -258,32 +260,40 @@ class WorkOrders
...
       else
         find_shaping_room(@hometown, @engineering_room)
       end
+      bput('swap', 'You move') if right_hand =~ /#{@noun}/i
 
-      wait_for_script_to_complete('carve', [recipe['chapter'], recipe['name'], info[stock_name], bone_carving ? 'stack' : 'rock', recipe['noun']])
+      wait_for_script_to_complete('carve', [recipe['chapter'], recipe['name'], info[stock_name], bone_carving ? 'stack' : material_noun[material_volume], recipe['noun']])
+      
+      material_volume = info['stock-volume'] if material_volume == 0
+      material_volume -= recipe['volume']
```
Comment (rpherbig) [inline workorders.lic]: Why subtract `recipe['volume']` here just to add it again on line 296?

File: workorders.lic (hunk 5 above)
Comment (Hazado) [reply workorders.lic]: Its hard to do that because of the loop its in

File: workorders.lic (hunk 6 above)
Comment (Hazado) [reply workorders.lic]: its not needed actually...the -=recipe['volume'] takes care of the issue.

File: workorders.lic (hunk 5 above)
Comment (Hazado) [reply workorders.lic]: Its not needed anymore.

**Comment by Hazado** [comment]
Any more issues with this anyone can see or have encountered?

**Comment by Tarhish** [comment]
Sorry, just poking around -- this is cool, but I had to run my crafter through its paces.

Knitting: seems to work as advertised.  Keeps yarn around and uses it.  I like this.

Sewing: does save the cloth but there's no logic to combine/count it, so it just stacks up burlap cloths in your bags.  We should probably either add counting/combining or remove retention for that discipline for now.

Carving stone:  Seems to work.  I can complete shaping and carving recipes, and although I don't have a good way to test work orders using pebbles to see if it uses their leftovers I think it looks right.

Shaping: it works, but when it scrapes lumber the lumber is dropped at your feet.  Previously, dispose() would handle picking up the lumber and dropping it again as trash, but since we're skipping dispose now it just sits there and relies on ;afk to pick it up before leaving the room.  Since ;afk is really more of a bandaid for a problem we intend to solve at some point, we should definitely handle stowing that lumber away.

**Comment by Hazado** [comment]
@Tarhish Please test again, should have fixed what you found.
Let me know how shaping goes, think the coding will work, but could use some improvement or fixes.

**Comment by Tarhish** [comment]
I can take a look at shaping -- I think it'll work better now.

For sewing, though, two things.  One, I would suggest adding stock name to the cloth whenever you use it.  There are going to be a lot of people out there with a chamois cloth in the same bag they use to craft, and that'll break everything.  Two, I think this will work to get the length of preexisting cloth just fine, but it doesn't combine them together.

I think just a get my burlap cloth and get my other burlap cloth, then combine and stow in the crafting bag will work.  The game response is "You combine the bolts of cloth together."  I'll try it out and see what I see.

**Comment by Tarhish** [comment]
Never mind, I'm totally wrong.  order_fabric will force a combine if your hands are full, so it should work.  Let me try it out as is

**Comment by Tarhish** [comment]
Sewing and Shaping both seem to do what's expected now.  I turned off afk to make sure I wouldn't miss some sleight of hand for foot-wood, and the cloths are being combined just fine.  I still would suggest adding the fabric type to the gets and counts though.  Most people will be running around with a cloth on their person, and it's best to avoid any possible ambiguity.

**Comment by Hazado** [comment]
Okay, made it more specific with the cloth section for sewing.
Try it one more time with new code and no issues, I think this is ready to be pushed.

**Comment by Tarhish** [comment]
Yeah, I don't see any problems with it.  I kept poking, but everything seems to be behaving itself.

**Comment by rpherbig** [comment]
Thank you @Hazado and @Tarhish !

**Pull request merged**